README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# devoutansi <img src="man/figures/logo.png" align="right" height=230/>

<!-- badges: start -->

![](https://img.shields.io/badge/cool-useless-green.svg)
![](http://img.shields.io/badge/dev-out-blue.svg) <!-- badges: end -->

`devoutansi` is a graphics output device for R which will output
graphics as text files with ANSI colouring.

## Installation

You can install from
[GitHub](https://github.com/coolbutuseless/devoutansi) with:

``` r
# install.packages("remotes")
remotes::install_github("coolbutuseless/devout")     # devout framework
remotes::install_github("coolbutuseless/miniansi")   # ANSI doc creator
remotes::install_github("coolbutuseless/devoutansi") # this device
```

#### Limitations

  - No current support for clipping paths or rasters.
  - Text positioning and spacing will always be a little off
  - Rotated text (except for 0 and 90 degrees) is probably going to be
    placed incorrectly.
  - No support for transparency i.e. alpha has no effect other then to
    suppress rendering completely when `alpha = 0`

# Basic Usage

Use `devoutansi::ansi()` as you would the `pdf()` or `png()` device.

The output will be written as ANSI characters to the screen, or a text
file if specified.

**Note:** - Because you can’t readily view ANSI in an HTML document, all
images included are screenshots of the output in my terminal.

**Note:** Click any image to view its ANSI text

``` r
library(ggplot2)
library(devoutansi)

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# PNG output of plot
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
p <- ggplot(mtcars) +
  geom_bar(aes(as.factor(cyl), fill = as.factor(cyl)), colour = NA) +
  theme_minimal() + 
  theme(
    panel.grid   = element_blank()
  ) +
  labs(
    title    = "{devoutansi} - ANSI graphics device",
    subtitle = "Hello #RStats"
  ) 

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# ANSI output
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
devoutansi::ansi(filename = "man/figures/mtcars1.txt", width = 120)
  p 
invisible(dev.off())
```

<a href="man/figures/mtcars1.txt"><img src="man/figures/mtcars1.png" width="100%" /></a>

All the demo plots on this page are available as their ANSI txt files by
clicking on the images.

## `geom_density()`

<a href="man/figures/mtcars3.txt"><img src="man/figures/mtcars3.png" width="100%" /></a>

<details>

<summary> Click to show/hide code </summary>

``` r
library(ggplot2)

p <- ggplot(mtcars) +
  geom_density(aes(mpg, fill = as.factor(cyl))) + 
  theme_minimal() +
  theme(
    panel.grid      = element_blank(), 
    legend.key.size = unit(5, 'in')
  ) + 
  scale_fill_viridis_d() + 
  labs(
    title    = "{devoutansi} - ANSI graphics device",
    subtitle = "filled polygons (geom_density)"
  )

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# ANSI output
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
devoutansi::ansi("man/figures/mtcars3.txt", width = 150)
p
invisible(dev.off())
```

</details>

## `geom_density()` in ASCII only

<a href="man/figures/mtcars3.txt"><img src="man/figures/mtcars3-ascii.png" width="100%" /></a>

<details>

<summary> Click to show/hide code </summary>

Extra options for ASCII mode:

  - `char_lookup_table` set to 1, 2 or 3 to select the character set
  - `pow` nonlinear intensity remapping using this exponent. default 1

<!-- end list -->

``` r
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# ANSI output
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
devoutansi::ansi("man/figures/mtcars3-ascii.txt", width = 150, plain_ascii = TRUE)
p
invisible(dev.off())
```

</details>

## `geom_sf()`

<a href="man/figures/geom-sf.txt"><img src="man/figures/geom-sf.png" width="100%" /></a>

<details>

<summary> Click to show/hide code </summary>

``` r
library(dplyr)
library(sf)
library(ggplot2)
library(devoutansi)

nc <- sf::st_read(system.file("shape/nc.shp", package = "sf"), quiet = TRUE)
nc <- nc %>% filter(between(CNTY_ID, 1820, 1830))

p <- ggplot(nc) +
  geom_sf(aes(fill = NAME)) +
  theme_bw(15) +
  theme(legend.key.size = unit(5, 'in')) + 
  scale_fill_brewer(palette = 'Dark2') + 
  labs(
    title    = "{devoutansi} - ANSI graphics device",
    subtitle = "filled polygons (geom_density)"
  )

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# ANSI output
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
devoutansi::ansi("man/figures/geom-sf.txt", width = 150)
p
invisible(dev.off())
```

</details>

## `pie` plot from base R

<a href="man/figures/pie.txt"><img src="man/figures/pie.png" width="100%" /></a>

<details>

<summary> Click to show/hide code </summary>

``` r
devoutansi::ansi("man/figures/pie.txt", width = 180)
pie(c(cool = 4, but = 2, use = 1, less = 8), main = "ANSI graphics device - Base pie plot")
invisible(dev.off())
```

</details>

## `geom_histogram()`

<a href="man/figures/mtcars2.txt"><img src="man/figures/mtcars2.png" width="100%" /></a>

<details>

<summary> Click to show/hide code </summary>

``` r
library(ggplot2)

p <- ggplot(mtcars) +
  geom_histogram(aes(mpg, fill = as.factor(cyl)), bins = 20) + 
  theme_minimal() +
  theme(
    panel.grid      = element_blank(),
    legend.key.size = unit(5, 'in')
  ) + 
  labs(
    title    = "{devoutansi} - ANSI graphics device",
    subtitle = "ggraph"
  ) + 
  scale_fill_brewer(palette = 'Set2')

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# ANSI output
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi("man/figures/mtcars2.txt", width = 150)
p
invisible(dev.off())
```

</details>

## `ggraph`

<a href="man/figures/ggraph.txt"><img src="man/figures/ggraph.png" width="100%" /></a>

<details>

<summary> Click to show/hide code </summary>

``` r
library(ggplot2)
library(ggraph)
library(tidygraph)
library(dplyr)

set.seed(1)
simple <- create_notable('bull') %>%
  mutate(name = c('But', 'Use', 'Cool', 'Less', 'Mike')) %>%
  activate(edges) %>%
  mutate(type = sample(c('friend', 'foe'), 5, TRUE))

p <- ggraph(simple, layout = 'graphopt') +
  geom_edge_link0(colour = 'grey70') +
  geom_node_text(aes(label = name), size = 10) + 
  theme_void() + 
  labs(
    title    = "{devoutansi} - ANSI graphics device",
    subtitle = "ggraph"
  )

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# ANSI output
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi("man/figures/ggraph.txt", width = 150)
p
invisible(dev.off())
```

</details>


NEWS.md
# devoutansi 0.1.0

* Initial release


R/ansi.R



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Graphics device for ANSI output
#'
#' Graphics primitives will be rendered using ANSI characters.
#'
#' Uses \code{devout::rdevice()}.
#'
#' @param filename If given, write ANSII to this file, otherwise write to console.
#' @param width,height dimensions of text output (in characters). Default: NULL
#'                     (auto-detect)
#' @param colour_depth ANSI colour depth. Default: 8.  8 is best for most terminals
#'        (including the Rstudio console).   24-bit depth is not supported by
#'        all terminals - use with caution.
#' @param plain_ascii Print ASCII representation only. default: FALSE
#' @param pow only valid when \code{plain_ascii = TRUE}. raise intensity to this
#'        power before conversion to an ASCII character.  Default: 1
#' @param char_lookup_table only valid when \code{plain_ascii = TRUE}. Choose from
#'        the available character mappings. Default: 1. Possible values 1, 2 or 3
#' @param font_aspect Character spacing horizontally and vertically are almost
#'        never the same. On many terminals character resolution vertically is
#'        half the resolution horizontally.  Adjust this value if circles don't
#'        look right. Default: 0.45
#' @param ... other parameters passed to the rdevice
#'
#' @import devout
#' @import miniansi
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi <- function(filename = NULL, width = NULL, height = NULL, colour_depth = 8,
                 plain_ascii = FALSE, pow = 1, char_lookup_table = 1, font_aspect = 0.45, ...) {
  devout::rdevice(ansi_callback, filename = filename, width = width, height = height,
                  colour_depth = colour_depth, plain_ascii = plain_ascii,
                  char_lookup_table = char_lookup_table, pow = pow,
                  font_aspect = font_aspect,
                  ..., device_name = 'ansi')
}


R/ansi-callback.R


`%||%` <- function (x, y) {
  if (is.null(x)) y else x
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' Convert a devices colour back into an RGB colour
#'
#' @param col 4 element RGBA colour with values in range [0-255]
#'
#' @importFrom grDevices rgb
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
col2hex <- function(col) {
  rgb(col[1], col[2], col[3], col[4], maxColorValue = 255)
}




#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# When the device is opened
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_open <- function(args, state) {

  width <- state$rdata$width
  if (is.null(width)) {
    width  <- getOption('width', default = 80)
  }

  height <- state$rdata$height
  if (is.null(height)) {
    # Set height to visually be half the width
    height <- as.integer(width * 0.5 * state$rdata$font_aspect)
  }

  state$rdata$width    <- width
  state$rdata$height   <- height
  state$dd$right       <- width   * 72
  state$dd$bottom      <- height  * 72
  state$dd$clipRight   <- width   * 72
  state$dd$clipBottom  <- height  * 72


  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # ANSI device
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  bg <- col2hex(state$dd$startfill)
  state$rdata$ansi <- miniansi::ANSI$new(width = width, height = height, background = bg, ansi_bits = state$rdata$colour_depth)

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # ASCII has different "inches per pixel" in the x and y direction.
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  font_aspect <- state$rdata$font_aspect %||% 0.45
  state$dd$ipr <- c(1/72, 1/72/font_aspect)
  state$dd$hasTextUTF8    <- TRUE  # The macOS terminal will let us use utf8
  state$dd$wantSymbolUTF8 <- TRUE



  state
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# When the device is closed
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_close <- function(args, state) {

  filename          <- state$rdata$filename
  ansi              <- state$rdata$ansi
  plain_ascii       <- isTRUE(state$rdata$plain_ascii)
  char_lookup_table <- state$rdata$char_lookup_table %||% 1
  pow               <- state$rdata$pow %||% 1

  if (!is.null(filename)) {
    ansi$save(filename, plain_ascii = plain_ascii, char_lookup_table = char_lookup_table, pow = pow)
  } else {
    ansi$print(plain_ascii = plain_ascii, char_lookup_table = char_lookup_table, pow = pow)
  }

  state
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Draw a line
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_line <- function(args, state) {

  x1 <- args$x1 / 72
  x2 <- args$x2 / 72
  y1 <- args$y1 / 72
  y2 <- args$y2 / 72

  state$rdata$ansi$line(x1, y1,  x2,  y2, col2hex(state$gc$col))

  state
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Draw a rectangle
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_rect <- function(args, state) {

  xmin <- round(min(c(args$x0, args$x1))/72)
  xmax <- round(max(c(args$x0, args$x1))/72)
  ymin <- round(min(c(args$y0, args$y1))/72)
  ymax <- round(max(c(args$y0, args$y1))/72)
  state$rdata$ansi$rect(
    xmin, ymin, xmax, ymax,
    colour = col2hex(state$gc$col),
    fill   = col2hex(state$gc$fill)
  )

  state
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Simply polyline
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_polyline <- function(args, state) {

  xs  <- as.integer(round(args$x / 72))
  ys  <- as.integer(round(args$y / 72))
  col <- col2hex(state$gc$col)

  state$rdata$ansi$polyline(xs, ys, colour = col)

  state
}




#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# No filled polygons (yet!)
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_polygon <- function(args, state) {

  xs   <- as.integer(round(args$x / 72))
  ys   <- as.integer(round(args$y / 72))
  col  <- col2hex(state$gc$col)
  fill <- col2hex(state$gc$fill)

  state$rdata$ansi$polygon(xs, ys, colour = col, fill = fill)

  state
}




#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Draw multiple paths
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_path <- function(args, state) {

  extents <- c(0, cumsum(args$nper))
  col     <- col2hex(state$gc$col)
  fill    <- col2hex(state$gc$fill)
  alpha   <- state$gc$fill[4]/255

  xs <- as.integer(round(args$x / 72))
  ys <- as.integer(round(args$y / 72))

  for (poly in seq_len(args$npoly)) {
    subargs   <- args
    lower     <- extents[poly     ] + 1L
    upper     <- extents[poly + 1L]
    x         <- xs[lower:upper]
    y         <- ys[lower:upper]
    state$rdata$ansi$polygon(
      xs          = x,
      ys          = y,
      colour      = col,
      fill        = fill
    )

  }


  state
}



#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Calculate and return the string width in the current state
#
# device_StrWidth should return the width of the given
# string in DEVICE units.
#
# graphics parameters that should be honoured (if possible):
#   font, cex, ps
#
# @param str string
#
# @return Optionally return 'width' the display width of the string in device units (numeric).
#         If not returned then a default value is used i.e. (strlen(str) + 2) * 72
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_strWidth <- function(args, state) {

  str <- args$str

  state$width <- (nchar(str) - 0.5) * 72

  state
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# For a text device, these numbers are pure fudge.
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_metricInfo <- function(args, state) {

  state$ascent  <- 0.5 * 72
  state$descent <- 0.7 * 72

  state
}


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# ref: http://members.chello.at/~easyfilter/bresenham.html
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_circle <- function(args, state) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Scales sizes and coords
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  r  <- ceiling(args$r/2)
  xc <- args$x / 72
  yc <- args$y / 72

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Draw a circle
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  state$rdata$ansi$circle(
    xc, yc, r = r,
    colour = col2hex(state$gc$col),
    fill   = col2hex(state$gc$fill)
  )

  state
}





#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# Text
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_text <- function(args, state) {
  x <- round(args$x/72)
  y <- round(args$y/72)


  str <- args$str

  # For text at 90degrees, y coord needs to be offset
  if (args$rot == 90) {
    n   <- nchar(str)
    n2  <- floor(n/2)
    y   <- y - n2
  }

  colour <- col2hex(state$gc$col)
  state$rdata$ansi$text(x, y, text = str, angle = args$rot, colour = colour)

  state
}






#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#' ASCII callback for the rdevice
#'
#' @param device_call name of device function call
#' @param args arguments to device function call
#' @param state list of rdata, dd and gc. Some or all of which may be NULL
#'
#' @export
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
ansi_callback <- function(device_call, args, state) {

  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # Anything we're not handling, just return() straight away
  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  # print(state$name)
  state <- switch(
    device_call,
    "open"         = ansi_open      (args, state),
    "close"        = ansi_close     (args, state),
    "line"         = ansi_line      (args, state),
    "polyline"     = ansi_polyline  (args, state),
    "circle"       = ansi_circle    (args, state),
    "rect"         = ansi_rect      (args, state),
    "text"         = ansi_text      (args, state),
    'strWidth'     = ansi_strWidth  (args, state),
    "textUTF8"     = ansi_text      (args, state),
    'strWidthUTF8' = ansi_strWidth  (args, state),
    'polygon'      = ansi_polygon   (args, state),
    'metricInfo'   = ansi_metricInfo(args, state),
    'path'         = ansi_path      (args, state),
    'circle'       = ansi_circle    (args, state),
    {
      # if (!device_call %in% c('strWidth', 'size', 'clip', 'mode', 'metricInfo')) {
      #   print(device_call);
      # }
      state
    }
  )

  state
}


